modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.transport;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.elasticsearch.common.SuppressForbidden;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+import static io.netty.channel.internal.ChannelUtils.MAX_BYTES_PER_GATHERING_WRITE_ATTEMPTED_LOW_THRESHOLD;
+
+
+/**
+ * This class is adapted from {@link NioSocketChannel} class in the Netty project. It overrides the channel
+ * read/write behavior to ensure that the bytes are always copied to a thread-local direct bytes buffer. This
+ * happens BEFORE the call to the Java {@link SocketChannel} is issued.
+ */
+@SuppressForbidden(reason = "Channel#write")
+public class CopyBytesSocketChannel extends NioSocketChannel {
+
+    private static final int MAX_BYTES_PER_WRITE = 1 << 20;
+
+    private static final ThreadLocal<ByteBuffer> ioBuffer = ThreadLocal.withInitial(() -> ByteBuffer.allocateDirect(MAX_BYTES_PER_WRITE));
+    private final WriteConfig writeConfig = new WriteConfig();
+
+    public CopyBytesSocketChannel() {
+        super();
+    }
+
+    CopyBytesSocketChannel(Channel parent, SocketChannel socket) {
+        super(parent, socket);
+    }
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
```

Comment:
Sorry for the delay here, I'll try my best to review this properly tomorrow. But one drive-by question:

If we do this, can't we get rid of the throttled/chunked writing I added to org.elasticsearch.transport.netty4.Netty4MessageChannelHandler#write in #39286 in case we use this channel implementation? (I suppose we'll still need it when we use the default channel, but it should give us a bit of speedup to not go through all the magic in that?)

Reply:
Maybe. We can discuss that as a follow up. There might still be value in that work to chunk TLS encoding (which is very CPU intensive). Enough value to extract some type of write limiter class that would also be added to the http transport.